src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/RenameTypeService.cs

```diff
@@ -0,0 +1,239 @@
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    [Export(typeof(IRenameTypeService))]
+    internal class RenameTypeService : IRenameTypeService
+    {
+        private readonly VisualStudioWorkspace _workspace;
+
+        [ImportingConstructor]
+        public RenameTypeService(
+            UnconfiguredProject project,
+            VisualStudioWorkspace visualStudioWorkspace)
+        {
+            _workspace = visualStudioWorkspace;
+            SyntaxFactsServicesImpl = new OrderPrecedenceImportCollection<ISyntaxFactsService>(projectCapabilityCheckProvider: project);
+        }
+
+        [ImportMany]
+        protected OrderPrecedenceImportCollection<ISyntaxFactsService> SyntaxFactsServicesImpl { get; }
+
+        private ISyntaxFactsService SyntaxFactsService
+        {
+            get
+            {
+                return SyntaxFactsServicesImpl.FirstOrDefault()?.Value;
+            }
+        }
+
+        public async Task<bool> AnyTypeToRenameAsync(string oldFilePath, string newFilePath, string projectPath)
+        {
+            string oldName = Path.GetFileNameWithoutExtension(oldFilePath);
+            string newName = Path.GetFileNameWithoutExtension(newFilePath);
+
+            if (!TryGetProjectAtPath(projectPath, out Project project))
+                return false;
+
+            if (!await CanHandleRenameAsync(oldName, newName, project))
+                return false;
+
+            (bool success, ISymbol symbol) = await TryGetSymbolToRenameAsync(oldName, newFilePath, project);
+            return success && symbol != null;
+        }
+
+        public async Task<bool> RenameTypeAsync(string oldFilePath, string newFilePath, string projectPath, CancellationToken cancellationToken)
+        {
+            string oldName = Path.GetFileNameWithoutExtension(oldFilePath);
+            string newName = Path.GetFileNameWithoutExtension(newFilePath);
+
+            if (!TryGetProjectAtPath(projectPath, out Project project))
+                return false;
+
+            if (!TryGetCodeModel(newFilePath, project, out EnvDTE.FileCodeModel codeModel))
+                return false;
+
+            (bool success, bool isCaseSensitive) = await TryDetermineIfCompilationIsCaseSensitiveAsync(project);
+            if (!success)
+                return false;
+
+            if (!TryGetCodeElementToRename(oldName, codeModel.CodeElements, isCaseSensitive, out EnvDTE80.CodeElement2 codeElementToRename))
+                return false;
+
+            codeElementToRename.RenameSymbol(newName);
+            return true;
+        }
+
+        private bool TryGetCodeModel(string newFilePath, Project project, out EnvDTE.FileCodeModel codeModel)
+        {
+            codeModel = null;
+            Document newDocument = GetDocument(project, newFilePath);
+            if (newDocument is null)
+                return false;
+
+            IVsHierarchy hierarchy = _workspace.GetHierarchy(project.Id);
```

Comment:
We have VisualStudioWorkspace.GetFileCodeModel() that takes a DocumentId and does all of this for you. (But see later comment which makes me wonder why you're using CodeModel at all.)

Reply:
Cos we need this code to run: https://github.com/dotnet/roslyn/blob/master/src/VisualStudio/Core/Impl/CodeModel/AbstractCodeModelService.cs#L497, in particular the refactor notifiers.

Reply:
Unless there's a better way, or a way to factor that code out into another service.

Reply:
@dpoeschl any idea what else can be done here?

Reply:
I talked to @dpoeschl 	in person.  There is no public API that handles rename tody while properly propagating all the VS events other than code model. If roslyn adds such an API I will happily consume it. I can file a bug on the roslyn repo about this if you want. The API would need to be MEF importable or on the VisualStudioWorkspace object I imagine.